thavalon-server/src/game/engine.rs

```diff
@@ -0,0 +1,761 @@
+        // TODO: there should probably be a "Going on a Mission" message
+        let game = self.game; // So we can use it inside the receive callback while mutably borrowing interactions
+        log::debug!("Sending {} on mission {}", proposal, mission);
+
+        let mut cards = HashMap::with_capacity(proposal.players.len());
+        while cards.len() != proposal.players.len() {
+            self.interactions
+                .receive(|player, action| {
+                    if proposal.players.contains(&player) {
+                        match action {
+                            Action::Play { card } => match cards.entry(player) {
+                                Entry::Occupied(entry) => {
+                                    Err(format!("You already played a {}", entry.get()))
+                                }
+                                Entry::Vacant(entry) => {
+                                    let role = game.players[player].role;
+                                    if role.can_play(card) {
+                                        entry.insert(card);
+                                        log::debug!("{} played a {}", game.name(player), card);
+                                        Ok(())
+                                    } else {
+                                        Err(format!("You can't play a {}", card))
+                                    }
+                                }
+                            },
+                            // TODO: fun on-mission actions
+                            _ => Err("You can't do that right now".to_string()),
+                        }
+                    } else {
+                        Err("You are not on the mission".to_string())
+                    }
+                })
+                .await?;
+        }
+
+        let failed = is_failure(game.spec, mission, cards.values());
+        log::debug!(
+            "Mission {} {}",
+            mission,
+            if failed { "failed" } else { "passed" }
+        );
+
+        self.interactions
+            .send(Message::MissionResults {
+                passed: !failed,
+                successes: cards.values().filter(|c| c == &&Card::Success).count(),
+                fails: cards.values().filter(|c| c == &&Card::Fail).count(),
+                reverses: cards.values().filter(|c| c == &&Card::Reverse).count(),
+            })
+            .await?;
+
+        if !failed {
+            let agravaine_declared = match time::timeout(
+                DECLARE_DELAY,
+                self.interactions.receive(|player, action| {
+                    let role = game.players[player].role;
+                    match action {
+                        Action::Declare if role == Role::Agravaine => Ok(true),
+                        _ => Err("You can't do that right now".to_string()),
+                    }
+                }),
+            )
+            .await
+            {
+                Ok(Ok(declared)) => declared,
```

Comment:
Where is this value coming from? Also, could we get rid of one of the `Ok`s by not returning `Ok(true)` on line 298, or is this separate?